### PR TITLE
Add basic exception-handling to Eval.

### PR DESCRIPTION
**Pull request by non**
These methods provide the basis for a `MonadError[Eval, Throwable]`
instance, without adding any overhead for people who don't wish to use
these (admittedly limited) capabilities.

Since `Eval` defers evaluation, we can take advantage of this to defer
additional recovery logic to the eventual evaluation. Critically, it doesn't
add any kind of *error node* to the `Eval` AST, or any extra overhead to
the default evaluation paths.

This came up when I was trying to catch up on https://github.com/typelevel/cats/pull/1552.
I realized there was no real reason why we couldn't have a
`MonadError[Eval, Throwable]` instance.

Review by @johnynek, @alexandru, et al.

### CODE REVIEW

**Comment by non** [comment]
I did a bunch of tests to convince myself that these methods preserve stack safety, and that they didn't consume additional stack unnecessarily.

(One exception I can think of is if the caller writes a `recovery` function which calls `.value` on the instance in question, and then nests these calls recursively. But this isn't something we can guard against easily and seems very unlikely in practice.)

File: tests/src/test/scala/cats/tests/EvalTests.scala [hunk 1]
```diff
@@ -86,7 +86,7 @@ class EvalTests extends CatsSuite {
     checkAll("Eval[Int]", MonadTests[Eval].monad[Int, Int, Int])
   }
   checkAll("Bimonad[Eval]", SerializableTests.serializable(Bimonad[Eval]))
-  checkAll("Monad[Eval]", SerializableTests.serializable(Monad[Eval]))
+  checkAll("MonadError[Eval, Throwable]", SerializableTests.serializable(MonadError[Eval, Throwable]))
```
Comment (johnynek) [inline tests/src/test/scala/cats/tests/EvalTests.scala]: this is only the serializable test, not the actual law test.

File: tests/src/test/scala/cats/tests/EvalTests.scala [hunk 2]
```diff
@@ -86,7 +86,7 @@ class EvalTests extends CatsSuite {
     checkAll("Eval[Int]", MonadTests[Eval].monad[Int, Int, Int])
```
Comment (johnynek) [inline tests/src/test/scala/cats/tests/EvalTests.scala]: note here is where the MonadTests are being run.

File: tests/src/test/scala/cats/tests/EvalTests.scala (hunk 1 above)
Comment (non) [inline tests/src/test/scala/cats/tests/EvalTests.scala]: Oh, good catch. I'll fix it now!

**Comment by johnynek** [comment]
👍 This tackles at least one use case @alexandru had.

**Comment by non** [comment]
I'm pushing some extra tests. Also, unfortunately I determined that my previous tests were wrong. The `.handleErrorWith` call does use one extra stack frame during `.value` evaluation. This is probably not a huge problem in most cases but does mean that if you nest e.g. 10K of these calls you'll hit a `StackOverflowException`.

I don't see a good way to remediating that issue without impacting the overall `Eval` performance. I'll push my extra tests and comments and we can decide then.

**Comment by non** [comment]
I see two sign-offs so I'm going ahead and merging. Thanks folks!

**Pull request merged**